blackbird_python/blackbird/program.py

```diff
@@ -314,25 +313,46 @@ def serialize(self):
 
         script = ["name {}".format(self.name), "version {}".format(self.version)]
 
-        if self.target["name"] is not None:
-            array_insert += 1
-            options = ""
+        # add target and type to the script
+        for name, data in [("target", self.target), ("type", self.programtype)]:
+            if data["name"] is not None:
+                array_insert += 1
+                options = ""
 
-            if self.target["options"]:
-                # if the target has options, compile them into
-                # the expected syntax
-                option_strings = [
-                    "{}={}".format(k, v) if not isinstance(v, str) else '{}="{}"'.format(k, v)
-                    for k, v in self.target["options"].items()
-                ]
-                options = " ({})".format(", ".join(option_strings))
+                if data["options"]:
+                    # if the target has options, compile them into
+                    # the expected syntax
+                    option_strings = [
+                        "{}={}".format(k, v) if not isinstance(v, str) else '{}="{}"'.format(k, v)
+                        for k, v in data["options"].items()
+                    ]
+                    options = " ({})".format(", ".join(option_strings))
 
-            # add target metadata
-            script.append("target {}{}".format(self.target["name"], options))
+                # add target metadata
+                script.append("{} {}{}".format(name, data["name"], options))
 
         # line break
         script.append("")
 
+        # add variables to the script
```

Comment:
Personally, would be in for separating parts of this method into multiple smaller methods (e.g., here having this as `add_variables`). (Commenting here, as this part of the method is newly added in this PR.)

Reply:
I fully agree with this. :100: I feel like this is the case for quite a few parts of the Blackbird code, so actually a refactor, that simply splits many parts of the code into sub-functions, would be really nice. Maybe this can be added separately as a refactor PR some time? What do you think @josh146?

Reply:
I agree! Definitely something that would be great to improve